src/mpd.c

```diff
@@ -679,17 +686,66 @@ append_string(char **a, const char *b, const char *separator)
   *a = temp;
 }
 
+/*
+ * Computes the absolute position of a relative position.  This is a
+ * feature introduced since MPD 0.23 where + or - for position can be
+ * used to indicate relative to the currently selected (playing/paused)
+ * song.
+ * This feature does the necessary lookups to resolve the current song
+ * and calculate the absolute position.  When ptype is POSITION_ABSOLUTE
+ * this function acts as a noop and simply returns position.
+ */
+static int
+mpd_get_relative_queue_pos(enum position_type ptype, int position)
+{
+  struct player_status status;
+  struct db_queue_item *queue_item;
+  uint32_t curpos;
+
+  /* shortcut absolute case */
+  if (ptype == POSITION_ABSOLUTE)
+    return position;
+
+  player_get_status(&status);
+
+  curpos = 0;
+  if (status.status != PLAY_STOPPED)
+    {
+      queue_item = db_queue_fetch_byitemid(status.item_id);
+      if (queue_item != NULL)
+  	{
+  	  if (queue_item->id > 0)
+    	    curpos = queue_item->pos;
+
+  	  free_queue_item(queue_item, 0);
+  	}
+    }
+
+  /* +0 inserts right after the current song */
+  if (ptype == POSITION_RELATIVE_AFTER)
+    position = curpos + position + 1;
+  else if (ptype == POSITION_RELATIVE_BEFORE)
+    position = curpos - position;
+
+  DPRINTF(E_DBG, L_MPD,
+      	  "current song: %d->%d, relative new position: %d\n",
+      	  status.item_id, curpos, position);
+
+  return position;
+}
+
 /*
  * Sets the filter (where clause) and the window (limit clause) in the given query_params
- * based on the given arguments
+ * based on the given arguments, returns position if requested and
+ * present
  *
  * @param argc Number of arguments in argv
  * @param argv Pointer to the first filter parameter
  * @param exact_match If true, creates filter for exact matches (e. g. find command) otherwise matches substrings (e. g. search command)
  * @param qp Query parameters
  */
 static int
-parse_filter_window_params(int argc, char **argv, bool exact_match, struct query_params *qp)
+parse_filter_window_params(int argc, char **argv, bool exact_match, struct query_params *qp, int *pos)
```

Comment:
This function is becoming a bit too complex, can you break it up? Also add the extra arg to the comment above. It would also be great if you could add some examples of real-world client requests to the comment.

Reply:
Sorry, but it still looks quite complex to me. Can you share examples of arguments incl expected query_params outcome, so I can better understand what the parser needs to do?

Reply:
this function is removed in its entirety in the refactor, sorry

Reply:
you'll want to look at https://github.com/owntone/owntone-server/pull/1793/commits/c54bb10663512005842a741925c2f32ac172043b#diff-725c859767cc777b1a9eafc699eea74762fd7120e4ca19385242210cf61bde48R948

Reply:
I did look at that, and I did see it's called mpd_parse_cmd_params(). That's the function that still looks quite complex. I must again ask you to share examples of input and output to understand what the parser needs to do.

Reply:
I'm affraid I need help as to what you find complex here.  The new function is just a loop over arguments, strcmp to compare each to a certain function, and then a switch-statement to execute the matching function.

On which of these three do you need more information?  The function is called by commands such as https://mpd.readthedocs.io/en/latest/protocol.html#command-addid and https://mpd.readthedocs.io/en/latest/protocol.html#command-playlistfind.  As you can see, `window`, `position` and so on appear in those, and they are the bits being parsed here.

Reply:
I didn't ask for an explanation of the function, I can read it just fine. I asked for examples of arguments (input), because I cannot guess that by myself. [Here](https://github.com/owntone/owntone-server/blob/master/src/httpd_rsp.c#L717) you can see from the RSP protocol what I mean.